In [1]:
import pandas as pd

In [2]:
df=pd.read_csv('marketing_data.csv')

In [3]:
df

,ID,Year_Birth,Education,Marital_Status,Income,Kidhome,Teenhome,Dt_Customer,Recency,MntWines,...,NumStorePurchases,NumWebVisitsMonth,AcceptedCmp3,AcceptedCmp4,AcceptedCmp5,AcceptedCmp1,AcceptedCmp2,Response,Complain,Country
0,1826,1970,Graduation,Divorced,84835.0,0,0,2014-06-16,0,189,...,6,1,0,0,0,0,0,1,0,Spain
1,1,1961,Graduation,Single,57091.0,0,0,2014-06-15,0,464,...,7,5,0,0,0,0,1,1,0,Canada
2,10476,1958,Graduation,Married,67267.0,0,1,2014-05-13,0,134,...,5,2,0,0,0,0,0,0,0,USA
3,1386,1967,Graduation,Together,32474.0,1,1,2014-05-11,0,10,...,2,7,0,0,0,0,0,0,0,Australia
4,5371,1989,Graduation,Single,21474.0,1,0,2014-04-08,0,6,...,2,7,1,0,0,0,0,1,0,Spain
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2235,10142,1976,PhD,Divorced,66476.0,0,1,2013-03-07,99,372,...,11,4,0,0,0,0,0,0,0,USA
2236,5263,1977,2n Cycle,Married,31056.0,1,0,2013-01-22,99,5,...,3,8,0,0,0,0,0,0,0,Spain
2237,22,1976,Graduation,Divorced,46310.0,1,0,2012-12-03,99,185,...,5,8,0,0,0,0,0,0,0,Spain
2238,528,1978,Graduation,Married,65819.0,0,0,2012-11-29,99,267,...,10,3,0,0,0,0,0,0,0,India


In [4]:
df.isnull().sum() #checking for null values 

ID                      0
Year_Birth              0
Education               0
Marital_Status          0
 Income                24
Kidhome                 0
Teenhome                0
Dt_Customer             0
Recency                 0
MntWines                0
MntFruits               0
MntMeatProducts         0
MntFishProducts         0
MntSweetProducts        0
MntGoldProds            0
NumDealsPurchases       0
NumWebPurchases         0
NumCatalogPurchases     0
NumStorePurchases       0
NumWebVisitsMonth       0
AcceptedCmp3            0
AcceptedCmp4            0
AcceptedCmp5            0
AcceptedCmp1            0
AcceptedCmp2            0
Response                0
Complain                0
Country                 0
dtype: int64

In [5]:
df.dtypes #only one column's data type should be corrected

ID                       int64
Year_Birth               int64
Education               object
Marital_Status          object
 Income                float64
Kidhome                  int64
Teenhome                 int64
Dt_Customer             object
Recency                  int64
MntWines                 int64
MntFruits                int64
MntMeatProducts          int64
MntFishProducts          int64
MntSweetProducts         int64
MntGoldProds             int64
NumDealsPurchases        int64
NumWebPurchases          int64
NumCatalogPurchases      int64
NumStorePurchases        int64
NumWebVisitsMonth        int64
AcceptedCmp3             int64
AcceptedCmp4             int64
AcceptedCmp5             int64
AcceptedCmp1             int64
AcceptedCmp2             int64
Response                 int64
Complain                 int64
Country                 object
dtype: object

In [6]:
df.columns = df.columns.str.strip()

In [7]:
df.columns # removed the unnecessary space behind and after income column

Index(['ID', 'Year_Birth', 'Education', 'Marital_Status', 'Income', 'Kidhome',
       'Teenhome', 'Dt_Customer', 'Recency', 'MntWines', 'MntFruits',
       'MntMeatProducts', 'MntFishProducts', 'MntSweetProducts',
       'MntGoldProds', 'NumDealsPurchases', 'NumWebPurchases',
       'NumCatalogPurchases', 'NumStorePurchases', 'NumWebVisitsMonth',
       'AcceptedCmp3', 'AcceptedCmp4', 'AcceptedCmp5', 'AcceptedCmp1',
       'AcceptedCmp2', 'Response', 'Complain', 'Country'],
      dtype='object')

In [8]:
df['Income']=df['Income'].fillna(df['Income'].median()) #fixing nulls

In [9]:
df.isnull().sum() 

ID                     0
Year_Birth             0
Education              0
Marital_Status         0
Income                 0
Kidhome                0
Teenhome               0
Dt_Customer            0
Recency                0
MntWines               0
MntFruits              0
MntMeatProducts        0
MntFishProducts        0
MntSweetProducts       0
MntGoldProds           0
NumDealsPurchases      0
NumWebPurchases        0
NumCatalogPurchases    0
NumStorePurchases      0
NumWebVisitsMonth      0
AcceptedCmp3           0
AcceptedCmp4           0
AcceptedCmp5           0
AcceptedCmp1           0
AcceptedCmp2           0
Response               0
Complain               0
Country                0
dtype: int64

In [10]:
df['Dt_Customer']=pd.to_datetime(df['Dt_Customer'])

In [11]:
df.dtypes

ID                              int64
Year_Birth                      int64
Education                      object
Marital_Status                 object
Income                        float64
Kidhome                         int64
Teenhome                        int64
Dt_Customer            datetime64[ns]
Recency                         int64
MntWines                        int64
MntFruits                       int64
MntMeatProducts                 int64
MntFishProducts                 int64
MntSweetProducts                int64
MntGoldProds                    int64
NumDealsPurchases               int64
NumWebPurchases                 int64
NumCatalogPurchases             int64
NumStorePurchases               int64
NumWebVisitsMonth               int64
AcceptedCmp3                    int64
AcceptedCmp4                    int64
AcceptedCmp5                    int64
AcceptedCmp1                    int64
AcceptedCmp2                    int64
Response                        int64
Complain    

In [12]:
print(df['AcceptedCmp1'].value_counts(normalize=True)*100)# percentage of people interacting with campaigns 0 means no 1 means yes
print(df['AcceptedCmp2'].value_counts(normalize=True)*100)
print(df['AcceptedCmp3'].value_counts(normalize=True)*100)
print(df['AcceptedCmp4'].value_counts(normalize=True)*100)
print(df['AcceptedCmp5'].value_counts(normalize=True)*100)
print(df['Response'].value_counts(normalize=True)*100)

AcceptedCmp1
0    93.571429
1     6.428571
Name: proportion, dtype: float64
AcceptedCmp2
0    98.660714
1     1.339286
Name: proportion, dtype: float64
AcceptedCmp3
0    92.723214
1     7.276786
Name: proportion, dtype: float64
AcceptedCmp4
0    92.544643
1     7.455357
Name: proportion, dtype: float64
AcceptedCmp5
0    92.723214
1     7.276786
Name: proportion, dtype: float64
Response
0    85.089286
1    14.910714
Name: proportion, dtype: float64


In [13]:
#utilizing sql to obtain answers of business questions, decomposing he tables further

In [14]:
customers=df[[ 'ID', 'Year_Birth', 'Education', 'Marital_Status', 'Income', 'Kidhome', 'Teenhome', 'Dt_Customer', 'Recency', 'Country', 'Complain']]

In [15]:
product_spending=df[[ 'ID', 'MntWines', 'MntFruits', 'MntMeatProducts', 'MntFishProducts', 'MntSweetProducts', 'MntGoldProds']]

In [16]:
purchase_channels=df[[ 'ID', 'NumDealsPurchases', 'NumWebPurchases', 'NumCatalogPurchases', 'NumStorePurchases', 'NumWebVisitsMonth']]

In [17]:
campaign_responses=df[['ID', 'AcceptedCmp1', 'AcceptedCmp2', 'AcceptedCmp3', 'AcceptedCmp4', 'AcceptedCmp5', 'Response']]

In [18]:
#dividing tables to load in sql

In [19]:
import sqlite3

In [20]:
con=sqlite3.connect('freshchoice.db')

In [21]:
customers.to_sql('customers', con, if_exists='replace', index=False)
product_spending.to_sql('product_spending', con, if_exists='replace', index=False)
purchase_channels.to_sql('purchase_channels', con, if_exists='replace', index=False)
campaign_responses.to_sql('campaign_responses', con, if_exists='replace', index=False)

2240

In [23]:
pd.read_sql("Select * from sqlite_master Where type= 'table'",con)

,type,name,tbl_name,rootpage,sql
0,table,customers,customers,2,"CREATE TABLE ""customers"" (\n""ID"" INTEGER,\n ""..."
1,table,product_spending,product_spending,3,"CREATE TABLE ""product_spending"" (\n""ID"" INTEGE..."
2,table,purchase_channels,purchase_channels,4,"CREATE TABLE ""purchase_channels"" (\n""ID"" INTEG..."
3,table,campaign_responses,campaign_responses,5,"CREATE TABLE ""campaign_responses"" (\n""ID"" INTE..."


In [39]:
pd.read_sql('''SELECT Response, ROUND(AVG(Income), 0) AS avg_income
                FROM customers
                JOIN campaign_responses 
                ON customers.ID = campaign_responses.ID
                WHERE Income IS NOT NULL
                GROUP BY Response;''', con)

,Response,avg_income
0,0,50846.0
1,1,60183.0


In [67]:
pd.read_sql('''SELECT Education, COUNT(*) AS count, SUM(Response) AS ACCEPTANCE
                FROM customers
                JOIN campaign_responses
                ON customers.ID=campaign_responses.ID
                GROUP BY Education''', con)

,Education,count,ACCEPTANCE
0,2n Cycle,203,22
1,Basic,54,2
2,Graduation,1127,152
3,Master,370,57
4,PhD,486,101


In [78]:
pd.read_sql('''SELECT Kidhome, SUM(cr.Response) AS accepted, COUNT(*) AS total
FROM customers c
JOIN campaign_responses cr ON c.ID = cr.ID
GROUP BY Kidhome;''', con)

,Kidhome,accepted,total
0,0,222,1293
1,1,110,899
2,2,2,48


In [79]:
pd.read_sql('''SELECT Kidhome, Teenhome, SUM(cr.Response) AS accepted, COUNT(*) AS total
FROM customers c
JOIN campaign_responses cr ON c.ID = cr.ID
GROUP BY Kidhome, Teenhome;''',con)

,Kidhome,Teenhome,accepted,total
0,0,0,169,638
1,0,1,50,625
2,0,2,3,30
3,1,0,66,503
4,1,1,42,374
5,1,2,2,22
6,2,0,2,17
7,2,1,0,31


In [80]:
pd.read_sql('''SELECT 
  CASE WHEN Income >= 80000 THEN 'High' ELSE 'Under 80k' END AS income_group,
  CASE WHEN Kidhome = 0 AND Teenhome = 0 THEN 'No Kids' ELSE 'Has Kids/Teens' END AS family_group,
  SUM(cr.Response) AS accepted,
  COUNT(*) AS total
FROM customers c
JOIN campaign_responses cr ON c.ID = cr.ID
GROUP BY income_group, family_group
ORDER BY accepted DESC;''',con)

,income_group,family_group,accepted,total
0,Under 80k,Has Kids/Teens,156,1573
1,Under 80k,No Kids,91,450
2,High,No Kids,78,188
3,High,Has Kids/Teens,9,29


In [81]:
#acceptance by country
pd.read_sql('''SELECT Country, COUNT(*) AS total, SUM(cr.Response) AS accepted
FROM customers c
JOIN campaign_responses cr ON c.ID = cr.ID
GROUP BY Country
ORDER BY accepted DESC;''', con)

Country,total,accepted
Spain,1095,176
Saudi Arabia,337,52
Canada,268,38
Australia,160,23
Germany,120,17
USA,109,13
India,148,13
Mexico,3,2


In [82]:
#average product spending by acceptance status
pd.read_sql('''SELECT cr.Response,
  ROUND(AVG(ps.MntWines), 0) AS avg_wines,
  ROUND(AVG(ps.MntFruits), 0) AS avg_fruits,
  ROUND(AVG(ps.MntMeatProducts), 0) AS avg_meat,
  ROUND(AVG(ps.MntFishProducts), 0) AS avg_fish,
  ROUND(AVG(ps.MntSweetProducts), 0) AS avg_sweets,
  ROUND(AVG(ps.MntGoldProds), 0) AS avg_gold
FROM campaign_responses cr
JOIN product_spending ps ON cr.ID = ps.ID
GROUP BY cr.Response;''', con)

Response,avg_wines,avg_fruits,avg_meat,avg_fish,avg_sweets,avg_gold
0,269.0,24.0,145.0,35.0,25.0,41.0
1,503.0,38.0,294.0,52.0,39.0,61.0


In [83]:
#average purchase channel usage by acceptance status
pd.read_sql('''SELECT cr.Response,
  ROUND(AVG(pc.NumDealsPurchases), 1) AS avg_deals,
  ROUND(AVG(pc.NumWebPurchases), 1) AS avg_web,
  ROUND(AVG(pc.NumCatalogPurchases), 1) AS avg_catalog,
  ROUND(AVG(pc.NumStorePurchases), 1) AS avg_store,
  ROUND(AVG(pc.NumWebVisitsMonth), 1) AS avg_web_visits
FROM campaign_responses cr
JOIN purchase_channels pc ON cr.ID = pc.ID
GROUP BY cr.Response;''', con)

Response,avg_deals,avg_web,avg_catalog,avg_store,avg_web_visits
0,2.3,3.9,2.4,5.7,5.3
1,2.3,5.1,4.2,6.1,5.3


In [84]:
#average previous campaign acceptance by campaign 6 response
pd.read_sql('''SELECT cr.Response,
  ROUND(AVG(cr2.AcceptedCmp1), 3) AS avg_cmp1,
  ROUND(AVG(cr2.AcceptedCmp2), 3) AS avg_cmp2,
  ROUND(AVG(cr2.AcceptedCmp3), 3) AS avg_cmp3,
  ROUND(AVG(cr2.AcceptedCmp4), 3) AS avg_cmp4,
  ROUND(AVG(cr2.AcceptedCmp5), 3) AS avg_cmp5
FROM campaign_responses cr
JOIN campaign_responses cr2 ON cr.ID = cr2.ID
GROUP BY cr.Response;''', con)

Response,avg_cmp1,avg_cmp2,avg_cmp3,avg_cmp4,avg_cmp5
0,0.034,0.005,0.045,0.055,0.037
1,0.237,0.060,0.231,0.186,0.275


## Analysis Summary
After cleaning and normalizing the data, SQL queries revealed some intresting insights for FreshChoice's Campaign 6 targeting:

1. **Income**: High-income customers (>80k) accept at 40% vs 12% for lower income brackets, showing income is a big factor
2. **Family Status**: Customers with no kids/teens accept at 26.5% compared to 10.3% for those with children - big difference
3. **Combined**: High-income + no kids has a 41.5% acceptance rate, thats nearly 3x the baseline of 14.9%
4. **Education**: PhD holders accept at 20.8% while Basic education only 3.7%
5. **Country**: Acceptance rates vary by region so geographic targeting could help
6. **Product Spend**: Accepters spend way more across all products specially wine and meat
7. **Channel**: Accepters prefer catalog and store purchases, non-accepters browse the web more but dont buy

**Recommendation**: Focus Campaign 6 on high-income customers without children - this 188-person segment gave 78 acceptances at 41.5% vs the overall 14.9% which is way better.